DirectProgramming/DPC++FPGA/ReferenceDesigns/cholesky/src/cholesky.hpp

```diff
@@ -0,0 +1,158 @@
+#ifndef __CHOLESKY_HPP__
+#define __CHOLESKY_HPP__
+
+#include <CL/sycl.hpp>
+#include <chrono>
+#include <cstring>
+#include <sycl/ext/intel/ac_types/ac_complex.hpp>
+#include <sycl/ext/intel/ac_types/ac_int.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <type_traits>
+#include <vector>
+
+#include "memory_transfers.hpp"
+#include "streaming_cholesky.hpp"
+#include "tuple.hpp"
+
+// Forward declare the kernel and pipe names
+// (This prevents unwanted name mangling in the optimization report.)
+class CholeskyDDRToLocalMem;
+class Cholesky;
+class CholeskyLocalMemToDDRL;
+class APipe;
+class QPipe;
+class RPipe;
+
+/*
+  Implementation of the Cholesky decomposition using multiple streaming kernels
+  Can be configured by datatype, matrix size (must use square matrices), real
+  and complex.
+*/
+template <unsigned columns,      // Number of columns in the input matrix
+          unsigned rows,         // Number of rows in the input matrix
+          unsigned raw_latency,  // RAW latency for triangular loop optimization
+          bool is_complex,       // Selects between ac_complex<T> and T datatype
+          typename T,            // The datatype for the computation
+          typename TT = std::conditional_t<is_complex, ac_complex<T>, T>
+          // TT will be ac_complex<T> or T depending on is_complex
+          >
+void CholeskyDecompositionImpl(
+    std::vector<TT> &a_matrix,  // Input matrix A to decompose
+    std::vector<TT> &l_matrix,  // Output matrix L
+    sycl::queue &q,             // Device queue
+    int matrix_count,           // Number of matrices to decompose
+    int repetitions             // Number of repetitions, for performance
+                                // evaluation
+) {
+  constexpr int kAMatrixSize = columns * rows;
+  constexpr int kLMatrixSize = columns * (columns + 1) / 2;
+  constexpr int kNumElementsPerDDRBurst = is_complex ? 4 : 8;
+
+  using PipeType = fpga_tools::NTuple<TT, kNumElementsPerDDRBurst>;
+
+  // Pipes to communicate the A and L matrices between kernels
+  using AMatrixPipe = sycl::ext::intel::pipe<APipe, PipeType, 3>;
+  using LMatrixPipe =
+      sycl::ext::intel::pipe<RPipe, TT, kNumElementsPerDDRBurst * 4>;
+
+  // Allocate FPGA DDR memory.
+  TT *a_device = sycl::malloc_device<TT>(kAMatrixSize * matrix_count, q);
+  TT *l_device = sycl::malloc_device<TT>(kLMatrixSize * matrix_count, q);
```

Comment:
Add checks for malloc_device returning nullptr and exit gracefully.

Reply:
OK - I should do the same for the QRD and QRI designs.